server/aesdsocket.c

```diff
@@ -193,155 +592,126 @@ void socketserver(int sck)
-                            if(errno == EINTR)
-                                continue;
-
-                            //Else, error occurred, print it to syslog and finish program
-                            syslog(LOG_ERR, "Could not read from the file: %s", strerror(errno));
-                            exit(1);
-                        }
-
-                        //Less bytes remaining
-                        to_be_sent -= read_bytes;
-                        
-                        //Send the contents back to the client
-                        int sent_bytes = -1;
-                        int send_off = 0;
-                        syslog(LOG_INFO, "Send bytes: %d", send_bytes);
-                        while(sent_bytes != 0)
-                        {
-                            sent_bytes = send(connection_fd, &buff_read[send_off], send_bytes, 0);
-                            if(sent_bytes == -1)
-                            {
-                                //If the error is caused by an interruption of the system call try again
-                                if(errno == EINTR)
-                                    continue;
-
-                                //Else, error occurred, print it to syslog and finish program
-                                syslog(LOG_ERR, "Could not read from the file: %s", strerror(errno));
-                                exit(1);
-                            }
-                            send_bytes -= sent_bytes;
-                            send_off += sent_bytes;
-                        }
-                        
-                    }
-
-                    //Reset index to use the malloc'ed buffer from the beginning
-                    index = 0;
-                }
-                //Realloc the array if it got full without an '\n'
-                else if(index == (RECV_BUFF_LEN*chunks))
-                {
-                    chunks++;
-                    recv_data = realloc(recv_data, sizeof(char)*RECV_BUFF_LEN*chunks);
-                    if(!recv_data)
-                    {
-                        syslog(LOG_ERR, "Could not realloc: %s", strerror(errno));
-                        exit(1);  
-                    }
+                    syslog(LOG_ERR, "Could not join thread: %s", strerror(ret));
+                    continue;  
                 }
+                //Free the memory used by the structure
+                free(element);
             }
+        }
+    }
 
-        } while(recv_ret != 0 && !graceful_exit);
+    ret = pthread_join(timestamp_thread, NULL);
+    if(ret != 0)
+    {
+        syslog(LOG_ERR, "Could not join thread: %s", strerror(ret));
+    }
 
-        //Free the used buffer
-        free(recv_data);
-        print_closed_conn(client_addr);
+exit_all:
+    if(sem_destroy(&sem_timestamp) == -1)
+    {
+        syslog(LOG_ERR, "Could not destroy semaphore: %s", strerror(errno));
+    }
+exit_filescktm:
+    ret = timer_delete(timer);
```

Comment:
@jordicrosmompart I'd recommend swapping the order of cleanup here, don't destroy the `sem_timestamp` until you delete the timer to prevent a case where the timer fires and attempts to access `sem_timestamp` after it's deleted.